### ISSUE TITLE
Lint for redundant cast 

### ISSUE THREAD

**Issue by malbarbo** [Manishearth/rust-clippy#1497]
How about linting things like:

```rust
pub fn g() -> u32 {
    10
}

pub fn f() {
    let _ = false as bool;
    let _ = 1u16 as u16;
    let _ = g() as u32;
}
```

**Comment by bood** [comment]
I can work on this one. Do we already have a module fit for this?

**Comment by oli-obk** [comment]
@bood: we have a `CastPass` https://github.com/Manishearth/rust-clippy/blob/37a0e52a1ed5077ab587f5524a9ac623cc71b223/clippy_lints/src/types.rs#L290

**Comment by bood** [comment]
@oli-obk Got it. Also:
1. Noticing all `CastPass` lints are `Allow`, should this one be `Allow` or `Warn`?
2. Should simply comparing `cast_from.sty == cast_to.sty` be enough here? Just not sure.

**Comment by oli-obk** [comment]
this one can be `Warn`, because it should not have any false positives.

Type equality can be funky. But only in the case that "equal" types might not compare equal (due to lifetimes). So that check should be ok.

**Comment by bood** [comment]
ok. BTW, which part of the `rustc::ty::TyS` specifies the lifetime of the type?
Looks like `TyS.sty` still equals for `x` `y` of different lifetime below:
```
let mut x = 1;
    {
        let mut y = 2;
        y = x;
    }
```

**Comment by oli-obk** [comment]
these types have no lifetime ;) They are plain integers. Only references have lifetimes. The allocations `x` and `y` have abstract lifetimes, but not the types of these allocations.
